File: provider/pkg/provider/certificate.go
package provider

import (
	"github.com/pulumi/pulumi-tls/sdk/v5/go/tls"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
)

type CertificateArgs struct {
	KeyPairArgs

	AllowedUses     pulumi.StringArrayInput        `pulumi:"allowedUses"`
	CaCertPem       pulumi.StringInput             `pulumi:"caCertPem"`
	CaPrivateKeyPem pulumi.StringInput             `pulumi:"caPrivateKeyPem"`
	IsCaCertificate pulumi.BoolInput               `pulumi:"isCaCertificate"`
	Subject         tls.CertRequestSubjectPtrInput `pulumi:"subject"`
}

type Certificate struct {
	KeyPair

	Cert *tls.LocallySignedCert `pulumi:"cert"`
	Csr  *tls.CertRequest       `pulumi:"csr"`
}

func NewCertificate(ctx *pulumi.Context,
	name string, args *CertificateArgs, opts ...pulumi.ResourceOption) (*Certificate, error) {
	if args == nil {
		args = &CertificateArgs{}
	}

	component := &Certificate{}
	err := ctx.RegisterComponentResource("kubernetes-the-hard-way:index:Certificate", name, component, opts...)
	if err != nil {
		return nil, err
	}

	key, err := tls.NewPrivateKey(ctx, name, &tls.PrivateKeyArgs{
		Algorithm:  args.Algorithm,
		EcdsaCurve: args.EcdsaCurve,
		RsaBits:    args.RsaBits,
	}, pulumi.Parent(component))
	if err != nil {
		return nil, err
	}

	csr, err := tls.NewCertRequest(ctx, name, &tls.CertRequestArgs{
		DnsNames:      args.DnsNames,
		IpAddresses:   args.IpAddresses,
		PrivateKeyPem: key.PrivateKeyPem,
		Subject:       args.Subject,
		Uris:          args.Uris,
	}, pulumi.Parent(component))

	cert, err := tls.NewLocallySignedCert(ctx, name, &tls.LocallySignedCertArgs{
		AllowedUses:         args.AllowedUses,
		CaCertPem:           args.CaCertPem,
		CaPrivateKeyPem:     args.CaPrivateKeyPem,
		CertRequestPem:      csr.PrivateKeyPem,
		EarlyRenewalHours:   args.EarlyRenewalHours,
		IsCaCertificate:     args.IsCaCertificate,
		SetSubjectKeyId:     args.SetSubjectKeyId,
		ValidityPeriodHours: args.ValidityPeriodHours,
	}, pulumi.Parent(component))
	if err != nil {
		return nil, err
	}

	component.Cert = cert
	component.CertPem = csr.CertRequestPem
	component.Csr = csr
	component.Key = key
	component.KeyPem = key.PrivateKeyPem

	if err := ctx.RegisterResourceOutputs(component, pulumi.Map{
		"cert":    cert,
		"certPem": csr.CertRequestPem,
		"csr":     csr,
		"key":     key,
		"keyPem":  key.PrivateKeyPem,
	}); err != nil {
		return nil, err
	}

	return component, nil
}

File: provider/cmd/pulumi-resource-kubernetes-the-hard-way/tls/certificate.ts
import { ComponentResourceOptions, Output, output } from '@pulumi/pulumi';
import { CertRequest, LocallySignedCert } from '@pulumi/tls';
import { CertRequestSubject } from '@pulumi/tls/types/output';
import * as schema from '../schema-types';
import { toAllowedUsage } from '../util';
import { File, InstallInputs, InstallOutputs, install } from '../remote';
import { KeyPair } from './keypair';

export class Certificate extends schema.Certificate {
  constructor(name: string, args: schema.CertificateArgs, opts?: ComponentResourceOptions) {
    super(name, args, opts);
    if (opts?.urn) return;

    const algorithm = output(args.algorithm ?? 'RSA');
    const allowedUses = output(args.allowedUses);
    const caCertPem = output(args.caCertPem);
    const caPrivateKeyPem = output(args.caPrivateKeyPem);
    const dnsNames = output(args.dnsNames ?? []);
    const ecdsaCurve = output(args.ecdsaCurve ?? 'P256');
    const ipAddresses = output(args.ipAddresses ?? []);
    const isCaCertificate = output(args.isCaCertificate ?? false);
    const subject = output(args.subject);
    const uris = output(args.uris ?? []);
    const validityPeriodHours = output(args.validityPeriodHours);

    const key = KeyPair.key(name, {
      algorithm,
      validityPeriodHours,
      ecdsaCurve,
      rsaBits: args.rsaBits,
    }, this);

    const csr = new CertRequest(name, {
      privateKeyPem: key.privateKeyPem,
      ipAddresses,
      dnsNames,
      uris,
      subject: args.subject,
    }, { parent: this });

    const cert = new LocallySignedCert(name, {
      isCaCertificate: args.isCaCertificate,
      allowedUses,
      validityPeriodHours,
      caCertPem,
      caPrivateKeyPem,
      certRequestPem: csr.certRequestPem,
      earlyRenewalHours: args.earlyRenewalHours,
      setSubjectKeyId: args.setSubjectKeyId
    }, { parent: this });

    this.algorithm = algorithm;
    this.allowedUses = cert.allowedUses.apply(toAllowedUsage);
    this.caCertPem = caCertPem;
    this.caPrivateKeyPem = caPrivateKeyPem;
    this.cert = cert;
    this.certPem = cert.certPem;
    this.csr = csr;
    this.dnsNames = dnsNames;
    this.earlyRenewalHours = cert.earlyRenewalHours;
    this.isCaCertificate = isCaCertificate;
    this.key = key;
    this.privateKeyPem = key.privateKeyPem;
    this.publicKeyPem = key.publicKeyPem;
    this.rsaBits = key.rsaBits;
    this.subject = subject as Output<CertRequestSubject> | undefined;
    this.uris = uris;
    this.validityPeriodHours = validityPeriodHours;

    this.registerOutputs({
      algorithm,
      allowedUses: this.allowedUses,
      caCertPem,
      caPrivateKeyPem,
      cert,
      certPem: this.certPem,
      csr,
      dnsNames,
      earlyRenewalHours: cert.earlyRenewalHours,
      isCaCertificate,
      key,
      privateKeyPem: this.privateKeyPem,
      publicKeyPem: this.publicKeyPem,
      rsaBits: this.rsaBits,
      subject: this.subject,
      uris,
      validityPeriodHours,
    });
  }

  public async installCert(args: InstallInputs): Promise<InstallOutputs> {
    return { result: installCert(this, args) };
  }

  public async installKey(args: InstallInputs): Promise<InstallOutputs> {
    return { result: installKey(this, args) };
  }
}

function installCert(cert: Certificate, args: InstallInputs): File {
  return install(args, cert.certPem);
}

function installKey(cert: Certificate, args: InstallInputs): File {
  return install(args, cert.publicKeyPem);
}
